pkg/client/kubernetes/clientmap/internal/shoot_clientmap.go

```diff
@@ -0,0 +1,123 @@
+	shootpkg "github.com/gardener/gardener/pkg/operation/shoot"
+
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	baseconfig "k8s.io/component-base/config"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type shootClientMap struct {
+	clientmap.ClientMap
+}
+
+func NewShootClientMap(factory *ShootClientSetFactory, logger logrus.FieldLogger) clientmap.ClientMap {
+	return &shootClientMap{
+		ClientMap: NewGenericClientMap(factory, logger),
+	}
+}
+
+type ShootClientSetFactory struct {
+	GetGardenClient        func(ctx context.Context) (kubernetes.Interface, error)
+	GetSeedClient          func(ctx context.Context, name string) (kubernetes.Interface, error)
+	ClientConnectionConfig baseconfig.ClientConnectionConfiguration
+
+	Log logrus.FieldLogger
+}
+
+func (f *ShootClientSetFactory) NewClientSet(ctx context.Context, k clientmap.ClientSetKey) (kubernetes.Interface, error) {
+	key, ok := k.(ShootClientSetKey)
+	if !ok {
+		return nil, fmt.Errorf("call to GetClient with unsupported ClientSetKey: expected %T got %T", ShootClientSetKey{}, k)
+	}
+
+	gardenClient, err := f.GetGardenClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get garden client: %w", err)
+	}
+
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := gardenClient.Client().Get(ctx, client.ObjectKey{Namespace: key.Namespace, Name: key.Name}, shoot); err != nil {
+		return nil, fmt.Errorf("failed to get Shoot object %q: %w", key.Key(), err)
+	}
+
+	if shoot.Spec.SeedName == nil {
+		return nil, fmt.Errorf("shoot %q is not scheduled yet", key.Key())
+	}
+
+	seedClient, err := f.GetSeedClient(ctx, *shoot.Spec.SeedName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get seed client: %w", err)
+	}
+
+	project, err := ProjectForNamespaceWithClient(ctx, gardenClient.Client(), shoot.Namespace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Project for Shoot %q: %w", key.Key(), err)
+	}
+
+	seedNamespace := shootpkg.ComputeTechnicalID(project.Name, shoot)
+
+	secretName := v1beta1constants.SecretNameGardener
+	// If the gardenlet runs in the same cluster like the API server of the shoot then use the internal kubeconfig
+	// and communicate internally. Otherwise, fall back to the "external" kubeconfig and communicate via the
+	// load balancer of the shoot API server.
+	addr, err := LookupHost(fmt.Sprintf("%s.%s.svc", v1beta1constants.DeploymentNameKubeAPIServer, seedNamespace))
+	if err != nil {
+		f.Log.Warnf("service DNS name lookup of kube-apiserver failed (%+v), falling back to external kubeconfig", err)
+	} else if len(addr) > 0 {
+		secretName = v1beta1constants.SecretNameGardenerInternal
+	}
+
+	clientOptions := client.Options{
+		Scheme: kubernetes.ShootScheme,
+	}
+
+	clientSet, err := NewClientFromSecret(ctx, seedClient, seedNamespace, secretName,
+		kubernetes.WithClientConnectionOptions(f.ClientConnectionConfig),
+		kubernetes.WithClientOptions(clientOptions),
+	)
```

Comment:
Why is this executed always?

Reply:
I am not sure, what you mean... 😅

Reply:
Sorry.. I meant: In line 113 you check if the `gardener-internal` secret can be used, so the lines 98-101 could have been put in the `else` block. But probably it was like this before and the clientset creation does not cause any "harm" in the form of I/O, so let's keep it like this. I'll resolve this comment.

Reply:
Yes, it was like this before, I just moved the code... :)